### PR TITLE
implementation of has_ipv4, has_any_ipv4, has_ipv4_prefix, and has_an…

### PR DESCRIPTION
**Pull request by bemitc**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added Kusto functions has_ipv4(), has_any_ipv4(), has_ipv4_prefix(), and has_any_ipv4_prefix()

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by kashwy — COMMENTED** [review]
1. Add unit test

2. Use `02366_kql` as the prefix for functional test files like other KQL functional test:
   02366_kql_func_dynamic.sql

3. Use  KQL statement in functional test.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return is_any ? 0 : 2; }
+    bool isVariadic() const override { return is_any ? true : false; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        auto argsLength = arguments.size();
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:40]: follow the style of variable name:
https://clickhouse.com/docs/en/development/style#names

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        return std::make_shared<DataTypeUInt8>();
+    }
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, const size_t input_rows_count) const override
+    {
+        auto argsLength = arguments.size();
+        std::vector<std::string> ipList;
+        auto bool_type = std::make_shared<DataTypeUInt8>();
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:87]: variable names

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        return std::make_shared<DataTypeUInt8>();
+    }
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, const size_t input_rows_count) const override
+    {
+        auto argsLength = arguments.size();
+        std::vector<std::string> ipList;
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:86]: https://clickhouse.com/docs/en/development/style#names
argsLength  , ipList

File: src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return is_any ? 0 : 2; }
+    bool isVariadic() const override { return is_any ? true : false; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        auto argsLength = arguments.size();
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp:39]: variable name

File: src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp [hunk 5]
```diff
@@ -284,4 +284,83 @@ bool FormatIpv4Mask::convertImpl(String & out, IParser::Pos & pos)
         kqlCallToExpression("ipv4_netmask_suffix", {"tostring(" + ip_address + ")"}, pos.max_depth));
     return true;
 }
+
+bool HasIpv4::convertImpl(String & out, IParser::Pos & pos)
+{
```
Comment (kashwy) [inline src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp:289]: can this use direct mapping?

File: src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp [hunk 6]
```diff
@@ -284,4 +284,83 @@ bool FormatIpv4Mask::convertImpl(String & out, IParser::Pos & pos)
...
+        out += std::format(
+            ",{0}",
+            ip);
+    }
+    out += ")";
+    return true;
 }
+
+bool HasIpv4Prefix::convertImpl(String & out, IParser::Pos & pos)
+{
```
Comment (kashwy) [inline src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp:328]: can this use direct mapping?

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp (hunk 1 above)
Comment (bemitc) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp]: I've removed the camelcase variable names

File: src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp (hunk 5 above)
Comment (bemitc) [reply src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp:289]: Yes. I've updated the non any functions to use direct mapping instead (which I was previously unaware of).

**Comment by ltrk2** [comment]
@bemitc in general, I think you could have more interesting functional tests. Feel free to experiment with edge cases (eg. empty strings, near-miss cases like '127.0.0.1111' or '127.0.0.a') and try to break your own implementation as much as you can. For starters, you could always include any test cases that the KQL specification of a given function contains.

I think you could also use just a single functional test case file, since the functions you are implementing are very much related and could benefit from the same test data. If you'd prefer to split it into multiple files though, that's also fine, just make sure to use the `02366_kql` prefix as @kashwy also mentioned.

File: tests/queries/0_stateless/02522_has_ipv4.sql [hunk 7]
```diff
@@ -0,0 +1,8 @@
+SELECT 'tests';
+DROP TABLE IF EXISTS has_ipv4_tbl;
+CREATE TABLE has_ipv4_tbl(ip String) ENGINE = Memory;
+INSERT INTO has_ipv4_tbl (ip) VALUES ('192.168.1.1');
+INSERT INTO has_ipv4_tbl (ip) VALUES ('127.0.0.1');
```
Comment (ltrk2) [inline tests/queries/0_stateless/02522_has_ipv4.sql:5]: You could perform these in a single query. The same goes for the other functional test files as well.

File: tests/queries/0_stateless/02523_has_any_ipv4.sql [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+INSERT INTO has_any_ipv4_str (ip) VALUES ('127.0.0.1');
+SELECT ip, kql_has_ipv4('XXXX 127.0.0.1', ip) from has_any_ipv4_str;
+DROP TABLE has_any_ipv4_str;
+DROP TABLE IF EXISTS has_any_ipv4_array;
+CREATE TABLE has_any_ipv4_array (ips Array(String)) ENGINE = Memory;
+INSERT INTO has_any_ipv4_array (ips) VALUES (['127.0.0.1', '127.0.0.2', '127.0.0.3']);
+INSERT INTO has_any_ipv4_array (ips) VALUES (['192.168.0.1', '192.168.0.2']);
+SELECT ips, kql_has_any_ipv4('XXX 127.0.0.3', ips) from has_any_ipv4_array;
+DROP TABLE has_any_ipv4_array;
+SELECT kql_has_any_ipv4('XXX 192.168.0.255', '192.168.0.255', '192.168.0.256');
```
Comment (ltrk2) [inline tests/queries/0_stateless/02523_has_any_ipv4.sql:14]: I think you could have the test data setup in the beginning and put the teardown at the very end and only have actual tests in between.

File: tests/queries/0_stateless/02522_has_ipv4.sql (hunk 7 above)
Comment (bemitc) [reply tests/queries/0_stateless/02522_has_ipv4.sql:5]: I had some problems with tests that throw exceptions. However, I see now you can add special formatted comments which indicate expected error codes which probably will help out. I'll take another pass on a better set of functional tests before I start working on unit tests.

File: src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp [hunk 9]
```diff
@@ -284,4 +284,63 @@ bool FormatIpv4Mask::convertImpl(String & out, IParser::Pos & pos)
...
+        first_address);
+
+    for (auto ip : ip_addrs)
+    {
+        out += std::format(
+            ",{0}",
+            ip);
+    }
+    out += ")";
+    return true;
```
Comment (ltrk2) [inline src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp:314]: I believe that you could mold the native function in a way that this preprocessing would no longer be necessary. Was there a particular reason that you chose this route?

Comment (bemitc) [reply src/Parsers/Kusto/KustoFunctions/KQLIPFunctions.cpp:314]: I missed the directMaping call when implementing convertImpl. It looks like I can directMap even variadic functions and let getReturnTypeImpl take care of enforcing the number of arguments (which it already does). I'll make that change.

**Review by ltrk2 — CHANGES_REQUESTED** [review]

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            {
+                const ColumnsWithTypeAndName isipv4string_args = {arguments[i]};
+                auto isipv4 = FunctionFactory::instance()
+                    .get("isIPv4String", context)
+                    ->build(isipv4string_args)
+                    ->execute(isipv4string_args, result_type, input_rows_count);
+                if (isipv4->getUInt(0) == 1)
+                {
+                    std::string ip = arguments[i].column->getDataAt(0).toString();
+                    ips.push_back(ip);
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:101]:
```suggestion
                    auto ip = arguments[i].column->getDataAt(0).toString();
                    ips.push_back(std::move(ip));
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                }
+            }
+        }
+
+        else if (isArray(arguments.at(1).type))
+        {
+            Field array0;
+            arguments[1].column->get(0, array0);
+            size_t len0 = array0.get<Array>().size();
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:110]:
```suggestion
            const auto len0 = array0.get<Array>().size();
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                    {
+                        ips.push_back(toString(array0.get<Array>().at(k)));
+                    }
+                }
+            }
+        }
+
+        if (!ips.empty())
+        {
+            std::string source = arguments[0].column->getDataAt(0).toString();
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:133]: You're treating argument #0 as if it could only ever be a literal. You need to perform this part for every row in the column and produce a result into a column of type `ColumnUInt8`.

Example, that works in ADX:
```
datatable (str:string) [
    'adsddsd a 127.0.0.1 u bn 192.168.0.1 aaa',
    'uuuuuu 255.255.255.255 weqee'
] | extend has_any_ipv4(str, '127.0.0.1');
```

In order to complicate things further, argument #1 can also be a column in case of `has_ipv4` as per ADX:
```
datatable (str:string, ip:string) [
    'adsddsd a 127.0.0.1 u bn 192.168.0.1 aaa', '127.0.0.1',
    'uuuuuu 255.255.255.255 weqee', '0.0.0.0'
] | extend has_ipv4(str, ip);
```

So you'll need a find a way to handle all of these.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+            std::regex ip_finder("([^[:alnum:]]|^)([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3})([^[:alnum:]]|$)");
+            std::smatch matches;
+            bool found_match = false;
+
+            if (std::regex_search(source, matches, ip_finder))
+            {
+                for (size_t i = 0; i < matches.size(); i++)
+                {
+                    if (std::any_of(ips.begin(), ips.end(), [i, matches](const std::string & str) -> bool { return str == matches[i]; }))
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:142]: It seems to me that `found_match` is unnecessary, since you can simply conclude any business once you find a match.

**Review by ltrk2 — CHANGES_REQUESTED** [review]
Please make similar changes to `FunctionKqlHasIpv4PrefixGeneric` than you will to `FunctionKqlHasIpv4Generic` once you go through all of my comments.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp (hunk 12 above)
Comment (bemitc) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp:133]: Hmm, i'll look into this, but in my testing I always seemed to be called one row at a time (with the first call when invoked from a table providing 0 rows).

image

Comment (ltrk2) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp:133]: It seems to me that it's a special case you've encountered: functions are executed on every part a table consists of. In your case, this happens because you're creating one row per part by performing inserts as separate queries. The same is true for every functional test you've written, so based on this it might seem that everything is fine. However, it isn't.

Consider the following example:
```
drop table if exists test;
create table test (str String, ip String) engine = Memory;
insert into test values ('adsddsd a 127.0.0.1 u bn 192.168.0.1 aaa', '127.0.0.1'), ('uuuuuu 255.255.255.255 weqee', '0.0.0.0');

test | extend has_ipv4(str, ip);
```

Result on your feature branch, since you're only considering the first row of any column:
```
┌─str──────────────────────────────────────┬─ip────────┬─Column1─┐
│ adsddsd a 127.0.0.1 u bn 192.168.0.1 aaa │ 127.0.0.1 │       1 │
│ uuuuuu 255.255.255.255 weqee             │ 0.0.0.0   │       1 │
└──────────────────────────────────────────┴───────────┴─────────┘
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        return std::make_shared<DataTypeUInt8>();
+    }
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, const size_t input_rows_count) const override
+    {
+        (void)input_rows_count;
+        const auto args_length = arguments.size();
+        auto result_column = ColumnUInt8::create();
+        const auto max_rows = std::max_element(arguments.begin(), arguments.end(), [] (ColumnWithTypeAndName const & lhs, ColumnWithTypeAndName const & rhs) { return lhs.column->size() < rhs.column->size(); })[0].column->size();
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:85]: why need this? there is already a` input_rows_count`

Comment (bemitc) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp:85]: Laszlo told me it was either 0 for a constant expression and 1 when it's a column, but that appears to perhaps no longer be the case. I'll remove this and make use of input_rows_count.

Comment (ltrk2) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp]: I still consider the issue Brian mentioned in his comment relevant. Please create functional tests that insert data in bulk in order to make sure that everything runs fine.

Comment (kashwy) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp]: input_rows_count is the one you need to handle the rows.

it can be 0, 1 or more than 1 , depends how the data is read  I think.

there are multiple threads to read data, so each thread can have 0 ,1 or more rows ,  this is passed to function as input_rows_count , and all args have the same rows, the previous code to get max size of args is not necessary.

to process the rows, iterate the rows from 0 to  input_rows_count is ok like:

```
for (size_t  i  = 0;  i < input_rows_count ; ++i )
{
  process each row
}
```

Comment (ltrk2) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp]: Hmm.. I see. Okay, sounds good to me!

Comment (bemitc) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp]: Sunday I did verify that input_rows_count contained the number of rows included in my insert (and stopped inserting 1 row at a time) before I made the change switching to input_rows_count instead of getting the maximum row count across all arguments. When I inserted 3 rows, input_rows_count was 3.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+        for (size_t i = 0; i < input_rows_count; i++)
+        {
+            bool res = false;
+            std::vector<std::string> ips;
+            if (isStringOrFixedString(arguments.at(1).type))
+            {
+                for (size_t j = 1; j < args_length; j++)
+                {
+                    auto arg = arguments[j].column->size() == input_rows_count ? arguments[j].column->getDataAt(i).toString() : arguments[j].column->getDataAt(0).toString();
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:90]: I don't understand why checking  `arguments[j].column->size() == input_rows_count`

Comment (bemitc) [reply src/Functions/Kusto/kqlHasAnyIpv4.cpp:90]: I'll change that. I had thought that a literal would have a size() of 1, but it looks like it's duplicated to match input_rows_count.

File: tests/queries/0_stateless/02366_kql_func_has_any_ipv4.sql [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+IP_STRING | extend has_ipv4(haystack, needle);
+IP_STRING | extend has_any_ipv4(haystack, needle);
+IP_STRING | extend has_any_ipv4(haystack, needle, '1.2.3.4');
+IP_STRING | extend has_any_ipv4(haystack, 'X', needle);
+IP_STRING | extend has_ipv4_prefix(haystack, needle);
+IP_STRING | extend has_ipv4_prefix(haystack, substring(needle, 0, strlen(needle)-1)); 
+IP_STRING | extend has_ipv4_prefix(haystack, substring(needle, 0, strlen(needle)-2));
+IP_ARRAY | extend has_any_ipv4(haystack, dynamic(needle));
+IP_ARRAY | extend has_any_ipv4_prefix(haystack, dynamic(needle));
```
Comment (ltrk2) [inline tests/queries/0_stateless/02366_kql_func_has_any_ipv4.sql:33]: You could have some kind of separation here to make it easier on someone trying to make a connection between the query and its reference output. For example: `print -- #1 --;` before each query with its unique ordinal.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+        return std::make_shared<DataTypeUInt8>();
+    }
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, const size_t input_rows_count) const override
+    {
+        const auto args_length = arguments.size();
+        auto result_column = ColumnUInt8::create();
+        auto isipv4string = [&, result_type] (ColumnsWithTypeAndName args) { return FunctionFactory::instance().get("isIPv4String", context)->build(args)->execute(args, result_type, 1); };
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:80]:
```suggestion
        const auto is_ipv4_string = [&, result_type] (ColumnsWithTypeAndName args) { return FunctionFactory::instance().get("isIPv4String", context)->build(args)->execute(args, result_type, 1); };
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+            if (isStringOrFixedString(arguments.at(1).type))
+            {
+                for (size_t j = 1; j < args_length; j++)
+                {
+                    auto arg = arguments[j].column->getDataAt(i).toString();
+
+                    ColumnPtr column_ip = DataTypeString().createColumnConst(1, toField(String(arg)));
+                    const ColumnsWithTypeAndName isipv4string_args = {ColumnWithTypeAndName(column_ip, std::make_shared<DataTypeString>(), "ip")};
+
+                    auto isipv4 = isipv4string(isipv4string_args);
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:95]:
```suggestion
                    const auto is_ipv4 = isipv4string(isipv4string_args);
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+            bool res = false;
+            std::vector<std::string> ips;
+            if (isStringOrFixedString(arguments.at(1).type))
+            {
+                for (size_t j = 1; j < args_length; j++)
+                {
+                    auto arg = arguments[j].column->getDataAt(i).toString();
+
+                    ColumnPtr column_ip = DataTypeString().createColumnConst(1, toField(String(arg)));
+                    const ColumnsWithTypeAndName isipv4string_args = {ColumnWithTypeAndName(column_ip, std::make_shared<DataTypeString>(), "ip")};
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:93]:
```suggestion
                    const ColumnsWithTypeAndName is_ipv4_string_args = {ColumnWithTypeAndName(column_ip, std::make_shared<DataTypeString>(), "ip")};
```

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+template <typename Name, bool is_any>
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:22]: Instead of the `is_any` parameter of type `bool` you could have something more descriptive, such as an `enum` (eg. `ArgumentPolicy` with choices like `Single` or `Variadic` or anything even more descriptive you can think of) that clearly explains the choice.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+                        const ColumnsWithTypeAndName isipv4string_args = {ColumnWithTypeAndName(column_ip, std::make_shared<DataTypeString>(), "ip")};
+
+                        auto isipv4 = isipv4string(isipv4string_args);
+                        if (isipv4->getUInt(0) == 1)
+                        {
+                            ips.push_back(toString(array0.get<Array>().at(j)));
+                        }
+                    }
+                }
+            }
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:123]: I think you could extract this into a separate function local to the translation unit, because its purpose is quite clear: to retrieve valid IP addresses from any arguments there may be.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+        auto result_column = ColumnUInt8::create();
+        auto isipv4string = [&, result_type] (ColumnsWithTypeAndName args) { return FunctionFactory::instance().get("isIPv4String", context)->build(args)->execute(args, result_type, 1); };
+
+        for (size_t i = 0; i < input_rows_count; i++)
+        {
+            bool res = false;
+            std::vector<std::string> ips;
+            if (isStringOrFixedString(arguments.at(1).type))
+            {
+                for (size_t j = 1; j < args_length; j++)
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:88]: If you wanted to make this part more elegant you could use [`std::ranges::copy_if`](https://en.cppreference.com/w/cpp/algorithm/ranges/copy) on the arguments with the first element dropped and the range transformed using `std::views::transform` with a lambda equivalent of `arguments[j].column->getDataAt(i).toString()` as the transform function. The predicate would essentially be the same as the inside of this scope, save for the `push_back`.

This comment also goes for the part where you do similar things to an array.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+            {
+                std::string source = arguments[0].column->getDataAt(i).toString();
+                std::regex ip_finder("([^[:alnum:]]|^)([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3})([^[:alnum:]]|$)");
+                std::smatch matches;
+
+                while (std::regex_search(source, matches, ip_finder))
+                {
+                    for (size_t j = 0; j < matches.size(); j++)
+                    {
+                        if (std::any_of(ips.begin(), ips.end(), [j, matches](const std::string & str) -> bool { return str == matches[j]; }))
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:135]: This should be a `find_if`, if I'm not mistaken. If so, you can use `std::ranges::find_if`.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+
+            if (!ips.empty())
+            {
+                std::string source = arguments[0].column->getDataAt(i).toString();
+                std::regex ip_finder("([^[:alnum:]]|^)([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3})([^[:alnum:]]|$)");
+                std::smatch matches;
+
+                while (std::regex_search(source, matches, ip_finder))
+                {
+                    for (size_t j = 0; j < matches.size(); j++)
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:133]: You could have `std::ranges::any_of` instead of this loop with a break and the `bool res`.

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+                        if (std::any_of(ips.begin(), ips.end(), [j, matches](const std::string & str) -> bool { return str == matches[j]; }))
+                        {
+                            res = true;
+                            break;
+                        }
+                    }
+                    source = matches.suffix().str();
+                }
+            }
+            result_column->insertValue(UInt8(res));
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:144]: In order to achieve something more performant at little cost in complexity you can save on virtual function calls by retrieving the underlying container using `ColumnUInt8::getData` and modifying that directly.

File: src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type of argument of function {}", getName());
+            }
+        }
+
+        else if (!is_any || !isArray(arguments.at(1).type))
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Illegal type of argument of function {}", getName());
+        }
+
+        return std::make_shared<DataTypeUInt8>();
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp:71]: This looks very similar to `FunctionKqlHasIpv4Generic`. If it's the same, you can just create a common header file and eliminate any duplication.

File: src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+template <typename Name, bool is_any>
+class FunctionKqlHasIpv4PrefixGeneric : public IFunction
```
Comment (ltrk2) [inline src/Functions/Kusto/kqlHasAnyIpv4Prefix.cpp:22]: Please try to apply my comments for `FunctionKqlHasIpv4Generic` to this one as well, because there's great similarity.

**Review by ltrk2 — CHANGES_REQUESTED** [review]

File: src/Functions/Kusto/kqlHasAnyIpv4.cpp [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, const size_t input_rows_count) const override
+    {
+        const auto args_length = arguments.size();
+        auto result_column = ColumnUInt8::create();
+        const auto isipv4string = [&, result_type] (ColumnsWithTypeAndName args) { return FunctionFactory::instance().get("isIPv4String", context)->build(args)->execute(args, result_type, 1); };
+
+        for (size_t i = 0; i < input_rows_count; i++)
+        {
```
Comment (kashwy) [inline src/Functions/Kusto/kqlHasAnyIpv4.cpp:83]: use `++i` instead of `i++`
https://clickhouse.com/docs/en/development/style
32. Always use the prefix increment/decrement operators if postfix is not required.

**Comment by kashwy** [comment]
@ltrk2 can you verify ? if all good , please re-base and merge. thanks

**Comment by ltrk2** [comment]
@kashwy sure, I'll take a look and discuss any remaining items with @bemitc.

**Review by ltrk2 — APPROVED** [review]
